src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs

```diff
@@ -368,179 +390,73 @@ internal abstract partial class AbstractInheritanceMarginService
-            Solution solution,
-            INamedTypeSymbol interfaceSymbol,
-            int lineNumber,
-            ImmutableArray<ISymbol> baseSymbols,
-            ImmutableArray<ISymbol> derivedTypesSymbols,
-            CancellationToken cancellationToken)
-        {
-            var baseSymbolItems = await baseSymbols
-                .SelectAsArray(symbol => symbol.OriginalDefinition)
-                .WhereAsArray(IsNavigableSymbol)
-                .Distinct()
-                .SelectAsArrayAsync((symbol, _) => CreateInheritanceItemAsync(
-                    solution,
-                    symbol,
-                    InheritanceRelationship.InheritedInterface,
-                    cancellationToken), cancellationToken)
-                .ConfigureAwait(false);
-
-            var derivedTypeItems = await derivedTypesSymbols
-                .SelectAsArray(symbol => symbol.OriginalDefinition)
-                .WhereAsArray(IsNavigableSymbol)
-                .Distinct()
-                .SelectAsArrayAsync((symbol, _) => CreateInheritanceItemAsync(solution,
-                    symbol,
-                    InheritanceRelationship.ImplementingType,
-                    cancellationToken), cancellationToken)
-                .ConfigureAwait(false);
-
-            return InheritanceMarginItem.CreateOrdered(
-                lineNumber,
-                topLevelDisplayText: null,
-                FindUsagesHelpers.GetDisplayParts(interfaceSymbol),
-                interfaceSymbol.GetGlyph(),
-                baseSymbolItems.Concat(derivedTypeItems));
-        }
-
-        private static async ValueTask<InheritanceMarginItem> CreateInheritanceMemberItemForInterfaceMemberAsync(
-            Solution solution,
+        private static InheritanceMarginItem CreateInheritanceMarginItem(
             ISymbol memberSymbol,
             int lineNumber,
-            ImmutableArray<ISymbol> implementingMembers,
-            CancellationToken cancellationToken)
-        {
-            var implementedMemberItems = await implementingMembers
-                .SelectAsArray(symbol => symbol.OriginalDefinition)
-                .WhereAsArray(IsNavigableSymbol)
-                .Distinct()
-                .SelectAsArrayAsync((symbol, _) => CreateInheritanceItemAsync(
-                    solution,
-                    symbol,
-                    InheritanceRelationship.ImplementingMember,
-                    cancellationToken), cancellationToken).ConfigureAwait(false);
-
-            return InheritanceMarginItem.CreateOrdered(
+            ImmutableArray<InheritanceTargetItem> targetItems)
+            => InheritanceMarginItem.CreateOrdered(
                 lineNumber,
                 topLevelDisplayText: null,
                 FindUsagesHelpers.GetDisplayParts(memberSymbol),
                 memberSymbol.GetGlyph(),
-                implementedMemberItems);
-        }
+                targetItems);
```

Comment:
if targetItems is empty, return null here (or potentially in CreateOrdered?)?

Reply:
Addressed